File: 2022/08/solution.pas
{ solution.pas -- treetop tree house -- Troy Brumley blametroi@gmail.com }

program solution;
{$longstrings off}
{$booleval on}


{ advent of code 2022 day 08 treetop tree house. given an 'orchard'
  of trees of varying heights, count trees based on their visibility
  or the quality of their scenic views. }


type
   integer = int64;

const
   MAXSIDE = 99;            { observed max }

{$I max.p }
{$I aocinput.p }


{ global variables are fine for this. the orchard is mapped into an
  array with some padding to simplify checking edge cases. }
var
   orchard    : array[0..MAXSIDE+1, 0..MAXSIDE+1] of char;
   sidelen    : integer;
   numvisible : integer;


{ initialize global variables to empty. orchard has extra rows
  and columns around the border to simplify visibility check. }

procedure init;

var
  i, j : integer;

begin
   for i := 0 to MAXSIDE+1 do
      for j := 0 to MAXSIDE+1 do
         orchard[i, j] := chr(0);
   sidelen := 0;
   numvisible := 0;
end;


{ check to see if a tree is visible from the four cardinal directions. }

function visible(r : integer;
                 c : integer) : boolean;
var
   i, j       : integer;
   t          : char;
   directions : integer;

begin
   directions := 4;
   t := orchard[r, c];
   for i := 0 to r-1 do
      if orchard[i, c] >= t then begin
         directions := directions - 1;
         break;
      end;
   { optimization here would be if was visible then done }
   for i := sidelen+1 downto r+1 do
      if orchard[i, c] >= t then begin
         directions := directions - 1;
         break;
      end;
   for j := 0 to c-1 do
      if orchard[r, j] >= t then begin
         directions := directions - 1;
         break;
      end;
   for j := sidelen+1 downto c+1 do
      if orchard[r, j] >= t then begin
         directions := directions - 1;
         break;
      end;
   visible := directions > 0;
end;


{ for part two, calculate the scenic score as how many trees can be
  seen before the view is blocked by a tree. trees on the edges are
  immediately blocked at one. multiply the number of trees in each
  direction, what is the highest scenic score? }

function scenicscore(r : integer;
                     c : integer) : integer;
var
   i, j   : integer;
   sn, ss,
   se, sw : integer;
   t      : char;

begin
   t := orchard[r, c];
   sn := 0; ss := 0; se := 0; sw := 0;
   for i := r-1 downto 1 do begin
      sn := sn + 1;
      if orchard[i, c] >= t then
         break;
   end;
   for i := r+1 to sidelen do begin
      ss := ss + 1;
      if orchard[i, c] >= t then
         break;
   end;
   for j := c-1 downto 1 do begin
      sw := sw + 1;
      if orchard[r, j] >= t then
         break;
   end;
   for j := c+1 to sidelen do begin
      se := se + 1;
      if orchard[r, j] >= t then
         break;
   end;
   scenicscore := sn * ss * se * sw;
end;


{ load the orchard and then look for visible trees. }

function partone : integer;

var
   i, j : integer;

begin
   init;
   aocrewind;
   i := 1; j := 1;

   while not eof(input) do begin
      if eoln(input) then begin
         readln(input);
         i := i + 1; j := 1;
         continue;
      end;
      read(input, orchard[i, j]);
      j := j + 1;
   end;

   sidelen := i - 1;
   numvisible := 0;
   for i := 1 to sidelen do
      for j := 1 to sidelen do
         if visible(i, j) then
            numvisible := numvisible + 1;
   partone := numvisible;
end;


{ part one loaded the orchard, scan to find the highest possible
  'scenic score'. knowing that trees on the sides have at least one
  scenic score of 0, there's no need to process them. }

function parttwo : integer;

var
   i, j : integer;
   s    : integer;

begin
   s := 0;
   for i := 2 to sidelen-1 do
      for j := 2 to sidelen-1 do
         s := max(s, scenicscore(i, j));
   parttwo := s;
end;


{ the mainline driver for a day's solution. }

procedure mainline;

var
   i : integer;

begin
   aocopen;
   init;
   i := partone;
   writeln('part one answer : ', i);
   i := parttwo;
   writeln('part two answer : ', i);
   aocclose;
end;


{ minimal boot to mainline }

begin
  mainline;
end.

File: 2022/08/solution.f90
! solution.f90 -- 2022 day 08 -- treetop tree house -- troy brumley blametroi@gmail.com

! advent of code 2022 day 08 treetop tree house. given an 'orchard' of trees of
! varying heights, count trees based on their visibility or the quality of their
! scenic views.

program solution

   use iso_fortran_env, only: int64
   use aocinput, only: open_aoc_input, read_aoc_input, close_aoc_input, max_aoc_reclen, rewind_aoc_input

   implicit none

   integer, parameter            :: AOCIN = 10
   integer(kind=int64)           :: part_one, part_two  ! results
   integer                       :: i, j                ! work
   character(len=max_aoc_reclen) :: rec

   ! orchard plot is 99x99, adding a padding cell around the edges
   ! to ease the coding
   integer, parameter            :: MAXSIDE = 99        ! observed
   character(len=1), dimension(0:MAXSIDE + 1, 0:MAXSIDE + 1) :: orchard
   integer                       :: sidelen

   ! input file is argument 1
   call open_aoc_input(AOCIN)

   ! initialize
   rec = ""
   part_one = 0; part_two = 0
   orchard = " "

   ! load the orchard. the test dataset is smaller than the max but still square
   ! use the number of input rows to limit the scope of searches
   i = 0
   do while (read_aoc_input(AOCIN, rec))
      if (len_trim(rec) > MAXSIDE) then
         print *, "*** error *** invalid input length at record", i + 1, "'", trim(rec), "'"
         stop 1
      end if
      i = i + 1
      do j = 1, 99
         orchard(i, j) = rec(j:J)
      end do
   end do
   sidelen = i

   ! for part one, how many trees can see to the edge of the orchard (can see
   ! the edge along at least one of the cardinal directions)

   do i = 1, sidelen
      do j = 1, sidelen
         if (visible(i, j)) part_one = part_one + 1
      end do
   end do

   ! for part two find the maximum scenic score (product of trees visible along
   ! the four cardinal directions)

   do i = 2, sidelen - 1
      do j = 2, sidelen - 1
         part_two = max(part_two, scenicscore(i, j))
      end do
   end do

   ! report and close
   print *
   print *, part_one, "part one"
   print *, part_two, "part two"
   print *

   call close_aoc_input(AOCIN)

contains

   ! is the tree at r, c visible from any of the four cardinal directions?
   function visible(ar, ac)
      implicit none
      integer, intent(in) :: ar, ac
      logical :: visible
      character(len=1) :: vt
      integer :: vd, vi, vj

      vd = 4;
      vt = orchard(ar, ac)

      do vi = 0, ar - 1
         if (orchard(vi, ac) >= vt) then
            vd = vd - 1
            exit
         end if
      end do
      ! we could exit after each of these checks if it didn't decrement
      ! vd, but it's cheap and clear to just do all

      do vi = sidelen + 1, ar + 1, -1
         if (orchard(vi, ac) >= vt) then
            vd = vd - 1
            exit
         end if
      end do

      do vj = 0, ac - 1
         if (orchard(ar, vj) >= vt) then
            vd = vd - 1
            exit
         end if
      end do

      do vj = sidelen + 1, ac + 1, -1
         if (orchard(ar, vj) >= vt) then
            vd = vd - 1
            exit
         end if
      end do

      visible = vd > 0
   end function visible

   ! for part two, calculate the senic score as how many trees can be seen
   ! before the view is blocked by another tree. trees on the edges are
   ! immediately blocked at one. multiple the number of trees in each cardinal
   ! direction and determine the highest scenic score.

   function scenicscore(ar, ac)
      implicit none
      integer, intent(in) :: ar, ac
      integer :: scenicscore
      integer :: vi, vj, vn, ve, vs, vw    ! subscripts, visible trees along direction
      character(len=1) :: vt               ! this tree

      vt = orchard(ar, ac)
      vn = 0; ve = 0; vs = 0; vw = 0

      do vi = ar - 1, 1, -1
         vn = vn + 1
         if (orchard(vi, ac) >= vt) exit
      end do

      do vi = ar + 1, sidelen
         vs = vs + 1
         if (orchard(vi, ac) >= vt) exit
      end do

      do vj = ac - 1, 1, -1
         vw = vw + 1
         if (orchard(ar, vj) >= vt) exit
      end do

      do vj = ac + 1, sidelen
         ve = ve + 1
         if (orchard(ar, vj) >= vt) exit
      end do

      scenicscore = vn*ve*vs*vw
   end function scenicscore

end program solution
